Build QNN in android_llm_demo.sh and perf

kirklandsign opened the pull request.

kirklandsign committed:
Build QNN in android_llm_demo.sh and perf

huydhn reviewed (approved):
LGTM!

huydhn commented on .github/workflows/android-perf.yml:
```diff
@@ -222,6 +222,11 @@ jobs:
         PYTHON_EXECUTABLE=python bash .ci/scripts/setup-linux.sh cmake
         export ARTIFACTS_DIR_NAME=artifacts-to-be-uploaded
 
+        if [[ ${{ matrix.delegate }} == "qnn" ]]; then
```
I just notice that the job fail https://github.com/pytorch/executorch/actions/runs/10713455099/job/29705837269#step:12:6342, you will need to define the matrix strategy for this job with

```
strategy:
  matrix:
    delegate: ${{ fromJson(needs.set-parameters.outputs.delegates) }}
fail-fast: false
```

kirklandsign replied:
Thank you!

guangy10 replied:
I just realized that we're building backend-specific apps in the android benchmark workflow and it seems like the changes are introduced here. @kirklandsign @huydhn I think we should not build different apps for each backend. There are several reasons for it:
  1. It will complicate the benchmarking workflow as you will have to add additional logics to match model with the app. Because we are benchmarking the perf of a model instead of the app, the app size is not a concern in this case.
  2. Delegates to hybrid backends will be supported for better perf at some point though today nobody is proactively working on it. Ideally QNN unsupported ops should fallback to XNNPACK instead of Portable, otherwise we may end up with QNN delegated model run slower than a pure CPU model w/ XNNPACK. Our stack should have supported building an app with multiple backends/delegates.

kirklandsign replied:
This is an addition to xnnpack when the QNN_SDK_ROOT is set 
(see https://github.com/pytorch/executorch/pull/5105/files#diff-20ef4ec16dbbb8f0311f92eb54751e47274a250335422f71a5107c4a053a1275R44-R45)

kirklandsign replied:
Right now we build this XNNPACK+QNN when matrix has qnn. https://github.com/pytorch/executorch/blob/main/.github/workflows/android-perf.yml#L147-L150

We *can* build a single flavor (XNNPACK+QNN+otherbackends) regardless of matrix as well

kirklandsign replied:
The requirement to do *always build QNN* is we need to run
```
PYTHON_EXECUTABLE=python bash .ci/scripts/setup-qnn-deps.sh
PYTHON_EXECUTABLE=python bash .ci/scripts/build-qnn-sdk.sh
```
all the time. Probably takes some time and stability. Ideally we have a docker image.

huydhn committed:
Update android-perf.yml to add matrix strategy to build-llm-demo job

kirklandsign commented:
Error: File /tmp/qnn/include/QNN/QnnTypes.h does not exist.

kirklandsign committed:
Update android-perf.yml

Should not export separately!
.ci/scripts/build-qnn-sdk.sh:  export QNN_SDK_ROOT=/tmp/qnn/2.23.0.240531

kirklandsign commented:
I'm not sure whether the error is from this PR or already there. 

https://github.com/pytorch/executorch/actions/runs/10715134533/job/29711462394
